### PR TITLE
JAMES-3774 cassandra driver 4

### PR DESCRIPTION
**Pull request by chibenwa**
TODO:

 - [x] Uncomment and update session instrumentation code
 - [x] Configuration options with the classpath method
 
 And of course...
 
  - [x] Propagate the changes throughout the application
      - [x] event-bus/cassandra
      - [x] event-sourcing/event-store-cassandra 
      - [x] mailbox/cassandra (**Quan**)
      - [x] mailbox/plugin/quota-mailing-cassandra
      - [x] server/blob/blob-cassandra (**Tung**)
      - [x] server/data/data-cassandra (**Tung**)
      - [x] server/data/data-jmap-cassandra (**Benoit**)
      - [x] server/protocols/protocols-pop3-distributed (**Benoit**)
      - [x] server/queue/queue-rabbitmq (Cassandra used for email query view)
      - [x] server/mailrepository/mailrepository-cassandra 
      - [x] server/task/task-distributed (**Benoit**)
      - [x] mailbox/plugin/deleted-messages-vault-cassandra (**Benoit**)
      - [x] Guice wiring....

### CODE REVIEW

**Review by quantranhong1999 — COMMENTED** [review]
Just read your PR. My IDE is showing compile error on `server/mailrepository/mailrepository-cassandra` so I will come for it first :')

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/ClusterFactory.java [hunk 1]
```diff
@@ -19,76 +19,47 @@
...
-        }
-    }
-
-    private static BoundStatement checkConnectionStatement(Session session) {
-        return session.prepare(select()
-                .fcall("NOW")
-                .from("local"))
-            .bind();
+    private static void ensureContactable(CqlSession session) {
+        session.execute("SELECT dateof(now()) FROM system.local ;");
```
Comment (quantranhong1999) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/ClusterFactory.java:63]:
```
[INFO] There is 1 error reported by Checkstyle 8.29 with checkstyle.xml ruleset.
[ERROR] src/main/java/org/apache/james/backends/cassandra/init/ClusterFactory.java:[62] (whitespace) EmptyLineSeparator: 'METHOD_DEF' should be separated from previous statement.
```

Please add a Empty line above L62

**Comment by quantranhong1999** [comment]
`server/mailrepository/mailrepository-cassandra` commit: https://github.com/quantranhong1999/james-project/commit/8ca179135b491f32e30efd923014c551eee21eee

Likely pick `mailbox/cassandra` next.

**Comment by vttranlina** [comment]
` server/data/data-cassandra` -> Tung

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraDefaultBucketDAO.java [hunk 2]
```diff
@@ -19,12 +19,11 @@
 
 package org.apache.james.blob.cassandra;
 
-import static com.datastax.driver.core.ConsistencyLevel.LOCAL_ONE;
-import static com.datastax.driver.core.querybuilder.QueryBuilder.bindMarker;
-import static com.datastax.driver.core.querybuilder.QueryBuilder.delete;
-import static com.datastax.driver.core.querybuilder.QueryBuilder.eq;
-import static com.datastax.driver.core.querybuilder.QueryBuilder.insertInto;
-import static com.datastax.driver.core.querybuilder.QueryBuilder.select;
+import static com.datastax.oss.driver.api.core.ConsistencyLevel.LOCAL_ONE;
```
Comment (chibenwa) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraDefaultBucketDAO.java:22]: We need to create an optimistic execution profile for this.

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java [hunk 3]
```diff
@@ -86,7 +97,7 @@ public Mono<byte[]> read(BlobId blobId) {
                 selectStatement.bind()
                     .setString(ID, blobId.asString())
                     .setConsistencyLevel(ONE)
-                    .setReadTimeoutMillis(readTimeOutFromDataBase))
+                    .setTimeout(Duration.ofDays(readTimeOutFromDataBase)))
```
Comment (chibenwa) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java:100]: Are we sure?

File: server/data/data-cassandra/src/main/java/org/apache/james/domainlist/cassandra/CassandraDomainList.java [hunk 4]
```diff
@@ -100,8 +89,8 @@ protected boolean containsDomainInternal(Domain domain) throws DomainListExcepti
     @Override
     public void addDomain(Domain domain) throws DomainListException {
         boolean executed = executor.executeReturnApplied(insertStatement.bind()
-            .setString(DOMAIN, domain.asString()))
-            .block();
+                .setString(DOMAIN, domain.asString()))
+            .blockOptional().orElseThrow();
```
Comment (chibenwa) [inline server/data/data-cassandra/src/main/java/org/apache/james/domainlist/cassandra/CassandraDomainList.java:93]: Why not `.block()` ?

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java (hunk 3 above)
Comment (vttranlina) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java:100]: My fault

File: server/data/data-cassandra/src/main/java/org/apache/james/domainlist/cassandra/CassandraDomainList.java (hunk 4 above)
Comment (vttranlina) [reply server/data/data-cassandra/src/main/java/org/apache/james/domainlist/cassandra/CassandraDomainList.java:93]: two way is the same, 
`block()` can be throw NullPointerException from Reactor (warning from IDE)
`blockOptional().orElseThrow()` for assert that

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraDefaultBucketDAO.java (hunk 2 above)
Comment (vttranlina) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraDefaultBucketDAO.java:22]: It's is old code, 
I only changed the import url
`import static com.datastax.driver.core.ConsistencyLevel.LOCAL_ONE;` 
-> 
`import static com.datastax.oss.driver.api.core.ConsistencyLevel.LOCAL_ONE;`

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java (hunk 3 above)
Comment (quantranhong1999) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java:100]: > My fault

Why is this a fault?

In my work, I even set a dedicated profile for this timeout. But turnout we have this `.setTimeout`.
Seem like Cassandra driver 4 does not have a dedicated timeout for read/write, it just has `REQUEST_TIMEOUT("basic.request.timeout")`.

Comment (vttranlina) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java:100]: Old: Millisecond 
My code: Day

It should be 

```suggestion
.setTimeout(Duration.ofMillis(readTimeOutFromDataBase)))
```
I got an issue with Cassandra query timeouts, I modified it for debug and forget to revert

Comment (quantranhong1999) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java:100]: Ah ok I got it.

Comment (chibenwa) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java:100]: To be set via a profile!

**Comment by vttranlina** [comment]
`server/queue/queue-rabbitmq (Cassandra used for email query view)` -> Tung

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/configuration/JamesExecutionProfiles.java [hunk 5]
```diff
@@ -27,6 +27,17 @@
 import com.datastax.oss.driver.api.core.config.DriverExecutionProfile;
 
 public interface JamesExecutionProfiles {
+    /**
+     * James application configuration allows for some operation to choose
+     * or not to enable LWTs.
+     *
+     * When LWTs are enable underlying DAOs will use LWT execution profiles.
+     */
+    enum ConsistencyChoice {
```
Comment (chibenwa) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/configuration/JamesExecutionProfiles.java:36]: How to use this?

```
class MyDAO {

   private final DriverExecutionProfile lwtProfile;
   
   MyDAO(CqlSession session) {
     // ... 
     lwtProfile = JamesExecutionProfiles.getLwtExecutionProfile(session);
   }

  //...
  
  public RESULT operation(Id id, ConsistencyChoice consistencyChoice) {
    BoundStatement boundStatement = statement.bind().setString(ID, id.serialize());
    if (consistencyChoice == STRONG) {
        boundStatement.setExecutionProfile(lwtProfile);
    }
    return executor.execute(boundStatement);
  }

```

Comment (quantranhong1999) [reply backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/configuration/JamesExecutionProfiles.java:36]: `setExecutionProfile` doc: 
```
All the driver's built-in implementations are immutable, and return a new instance from this method. However custom implementations may choose to be mutable and return the same instance.
```

I propose: 
```
private BoundStatement setExecutionProfileIfNeeded(BoundStatement statement, JamesExecutionProfiles.ConsistencyChoice consistencyChoice) {
        if (consistencyChoice.equals(STRONG)) {
            return statement.setExecutionProfile(lwtProfile);
        } else {
            return statement;
        }
    }
```

Comment (vttranlina) [reply backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/configuration/JamesExecutionProfiles.java:36]: @quantranhong1999  what is the difference?
BoundStatement is unique for each query?

**Comment by chibenwa** [comment]
Opened a JIRA ticket: https://issues.apache.org/jira/browse/JAMES-3774

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraCacheConfiguration.java [hunk 6]
```diff
@@ -33,25 +33,12 @@
...
-        private static final Duration DEFAULT_READ_TIMEOUT = Duration.ofMillis(100);
-        private static final Duration MAX_READ_TIMEOUT = Duration.ofHours(1);
         private static final Duration DEFAULT_TTL = Duration.ofDays(7);
         private static final int DEFAULT_BYTE_THRESHOLD_SIZE = 8 * 1024;
 
-        private Optional<Duration> readTimeout = Optional.empty();
         private Optional<Integer> sizeThresholdInBytes = Optional.empty();
         private Optional<Duration> ttl = Optional.empty();
 
-        public Builder timeOut(Duration timeout) {
```
Comment (vttranlina) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraCacheConfiguration.java:45]: Don't forget to remove `cache.cassandra.timeout`  in documents, if it is not used

**Comment by vttranlina** [comment]
## Note
Where invoke the `org.apache.james.backends.cassandra.StatementRecorder.Selector#preparedStatement`
We should remove the `;` in the query statement
Example:
```java
       assertThat(statementRecorder.listExecutedStatements(
            StatementRecorder.Selector.preparedStatement("SELECT domain FROM domains WHERE domain=:domain;")))
            .hasSize(1);
```
-> 
```java
       assertThat(statementRecorder.listExecutedStatements(
            StatementRecorder.Selector.preparedStatement("SELECT domain FROM domains WHERE domain=:domain")))
            .hasSize(1);
```

**Comment by chibenwa** [comment]
> Where invoke the org.apache.james.backends.cassandra.StatementRecorder.Selector#preparedStatement
> We should remove the ; in the query statement

And we should also put everyting in lower case ;-)

**Review by Arsnael — COMMENTED** [review]
Not much to comment for now...

Don't forget to add the ticket number on your commits at the end though :)

**Comment by vttranlina** [comment]
Should we carry about the naming strategy?
By default, naming strategy of all queries (create table, type, query table...) will be converted to lowercase. 
This come from `CqlIdentifier.fromCql`, 
```java
// from Cassandra lib
public static CqlIdentifier fromCql(@NonNull String cql) {
        Preconditions.checkNotNull(cql, "cql must not be null");
        String internal;
        if (Strings.isDoubleQuoted(cql)) {
            internal = Strings.unDoubleQuote(cql);
        } else {
            internal = cql.toLowerCase(Locale.ROOT);
            Preconditions.checkArgument(!Strings.needsDoubleQuotes(internal), "Invalid CQL form [%s]: needs double quotes", cql);
        }

        return fromInternal(internal);
    }
```

Is it a potential conflict table when deployed?

**Comment by chibenwa** [comment]
> Is it a potential conflict table when deployed?

No. All stuff on Cassandra side is lowercase. CqlIdentifier allows to reduce impact of always calling toLowerCase...

**Comment by chibenwa** [comment]
-> https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-599/32/testReport/junit/org.apache.james.queue.rabbitmq/RabbitMQMailQueueTest$MailQueueSizeMetricsEnabled/messagesShouldBeProcessedAfterNotPublishedMailsHaveBeenReprocessedAndNewMessagesShouldNotBeLost/
 
  -> Also worth it to have a look at MPT IMAP Cassandra stuff

**Comment by vttranlina** [comment]
When I run distributed container (`apache/james:distributed-latest`) ->  James start fail, it hangling 
But it work with the cassandra app container (`apache/james:cassandra-latest`). 

Docker-compose reference: server/apps/distributed-app/docker-compose.yml
Container log: 

image

// No any error with distributed integration test

**Comment by chibenwa** [comment]
Likely à missing cassandra.local.dc in cassandra.properties...

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**